### DISCUSSION TITLE
Errors in generateMetadata are not handled by error.tsx

### DISCUSSION THREAD

**Discussion by codinginflow** [vercel/next.js#49925] [App Router]
Often you need to fetch some data in `generateMetadata`:

```
export async function generateMetadata({ params: { username } }: PageProps): Promise<Metadata> {
    try {
        const user = await getUser(username);
        return {
            title: `${user.username} - Flow Blog`,
        };
    } catch (error) {
        if (error instanceof NotFoundError) {
            notFound();
        } else {
            throw error; 
        }
    }
}
```

**The problem**: The error we rethrow in the `catch` block doesn't lead to the `error.tsx` page. Instead, we get to a default full-screen error page that doesn't include any elements of our `layout.tsx`.

How am I supposed to make network requests in `generateMetadata` if errors are not handled properly?

**Comment by tom-sherman** [comment]
This really feels like a bug in Next.js.

**Reply by tom-sherman** [reply]
I feel strongly that this is a bug for dev mode so have raised it here https://github.com/vercel/next.js/issues/50863

**Comment by arjenbloemsma** [comment]
Hi @codinginflow,

To catch those errors you you need to add a 'global-error.tsx' to your app directory.

You can find more details in the nextjs docs [here](https://nextjs.org/docs/app/building-your-application/routing/error-handling#handling-errors-in-root-layouts).

Hope this helps.

Kind regards,

Arjen

**Reply by codinginflow** [reply]
But I would like to catch the errors on the level of the page where they happen.

**Reply by tom-sherman** [reply]
I don't think that's possible, pretty sure generateMetadata runs in parallel across all route segments.

**Reply by codinginflow** [reply]
> I don't think that's possible, pretty sure generateMetadata runs in parallel across all route segments.

What is the correct thing to do with errors in `generateMetadata` then? Should I return an empty object?

**Reply by tom-sherman** [reply]
That depends on your specific use case. Returning an empty object is probably a good first pass but in other cases you might want to return partial metadata in a best-effort manner.

**Comment by p00000001** [comment]
I have a similar issue with data fetching and calling `notFound()` from `generateMetadata()` that it skips the `not-found.js` page and defaults to `global-error.js`, I raised as an issue in #52329 

I have worked around it by catching errors in `generateMetadata()` and returning blank metadata, and then the `notFound()` error throws again in the default page function.

However, I do agree that it seems more intuitive for `generateMetadata()` to handle errors in the same way as the default page function. This simplifies the code, and is likely a faster response because  `generateMetadata()` typically returns first.

**Comment by hristokoev** [comment]
I'm now on Nextjs 15.0.0-canary.104 and this appears to be an ongoing issue. Errors from within the `generateMetadata` are not being redirected to the `error.tsx` that I've set up. Interestingly enough, on dev this works, but in production it doesn't.

**Reply by alvarlagerlof** [reply]
Please consider making a reproduction and open an issue. Considering https://github.com/vercel/next.js/issues/50863 was already resolved, I doubt anyone will look here unless there is a new issue opened.
